### ISSUE TITLE
Optimize split based on target

### ISSUE THREAD

**Issue by thegoldgoat** [thegoldgoat/cicero_compiler_cpp#9]
If split target is a jump, delete that jump and update the target of the split with the target of the jump

For example, for regex `^x?`

```
module {
  "cicero.flat_split"() {splitTarget = @FLATTEN_0} : () -> ()
  "cicero.match_char"() {targetChar = 120 : i8} : () -> ()
  "cicero.accept_partial"() {sym_name = "S0"} : () -> ()
  "cicero.jump"() {sym_name = "FLATTEN_0", target = @S0} : () -> ()
}
```

While we could have something like

```
module {
  "cicero.flat_split"() {splitTarget = @S0} : () -> ()
  "cicero.match_char"() {targetChar = 120 : i8} : () -> ()
  "cicero.accept_partial"() {sym_name = "S0"} : () -> ()
}
```

In this case, another optimization could be done: we are only interested in the matching, so in this case the target of the split is an accept/accept_partial, which means that it always accepts. The difficulty here is that it is not as simple as eliminating the flat split, we also need to remove all unreachable code blocks, for example doing just this is enough but not optimal:

```
module {
  "cicero.accept_partial"() : () -> ()
  "cicero.match_char"() {targetChar = 120 : i8} : () -> ()
  "cicero.accept_partial"() {sym_name = "S0"} : () -> ()
}
```

Because `match_char` and the second `accept_partial` are not reachable anymore and should be removed! Is it worth implementing control flow analysis?

**Comment by thegoldgoat** [comment]
Also, consider solving this problem by optimizing regex MLIR, for example:

```
❯ ./ciceroc --regex="^x?" --emit=regexmlir -Oall
"builtin.module"() ({
  "regex.root"() ({
    "regex.concatenation"() ({
      "regex.piece"() ({
        "regex.match_char"() {targetChar = 120 : i8} : () -> ()
        "regex.quantifier"() {max = 1 : si64, min = 0 : si64} : () -> ()
      }) : () -> ()
    }) : () -> ()
  }) {hasPrefix = false, hasSuffix = true} : () -> ()
}) : () -> ()
```

In this case, given we have `hasSuffix = true` we can simplify the whole `x?` with an `accept_partial`

**Comment by thegoldgoat** [comment]
Implemented in f47c376ccc4f80b787e4660a46ffb7f1e2e137d0 as a middle-end optimization on regex MLIR
